### PR TITLE
rocks_engine impl snapshot.Base

### PR DESCRIPTION
**Pull request by wenbobuaa**
# Description           

snapshot.rocks_engine 实现 Base trait

## Type of change       

- **New feature**       

# Checklist:

- [x] **Self-review**: I have performed a **self-review** of my own code
- [ ] **Comment**:     I have **commented my code**, particularly in hard-to-understand areas
- [ ] **Doc**:         I have made corresponding changes to the **documentation**
- [x] **No-warnings**: My changes generate **no new warnings**
- [x] **Add-test**:    I have added **tests** that prove my fix is effective or that my feature works
- [x] **Pass**:        New and existing **unit tests pass** locally with my changes

### CODE REVIEW

File: components/epaxos/src/snapshot/rocks_engine/test_engine.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    let next0 = eng.next_kv(&"k0".as_bytes().to_vec(), true);
+    assert_eq!(
+        next0,
+        Some(("k0".as_bytes().to_vec(), "v0".as_bytes().to_vec()))
+    );
+
+    let next1 = eng.next_kv(&"k0".as_bytes().to_vec(), false);
+    assert_eq!(
+        next1,
+        Some(("k1".as_bytes().to_vec(), "v1".as_bytes().to_vec()))
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/rocks_engine/test_engine.rs]: 这里直接引用kvs里的元素看着能简单点好像

File: components/epaxos/src/snapshot/rocks_engine/engine.rs [hunk 2]
```diff
@@ -125,6 +126,50 @@ impl RocksDBEngine {
...
+    fn next_kv(&self, key: &Vec<u8>, include: bool) -> Option<(Vec<u8>, Vec<u8>)> {
+        let mut iter = self.db.iter();
+
+        iter.seek(SeekKey::from(&key[..]));
+
+        while iter.valid() {
+            match iter.kv() {
+                Some(kv) => {
+                    if key == &kv.0 && !include {
+                        iter.next();
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/rocks_engine/engine.rs]: 这块先判断下include, 去掉while循环感觉好点

File: components/epaxos/src/snapshot/rocks_engine/mod.rs [hunk 3]
```diff
@@ -7,6 +7,9 @@ use rocks::*;
 mod engine;
 pub use engine::*;
 
+#[cfg(test)]
+mod test_engine;
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/rocks_engine/mod.rs]: 这块用上层的那个test_engine吧, 需要跟mem_engine共享一些代码.

File: components/epaxos/src/snapshot/rocks_engine/test_engine.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    let iter = eng.get_iter("k1".as_bytes().to_vec(), true);
+    for (idx, item) in iter.enumerate() {
+        assert_eq!(kvs[idx], item)
+    }
+}
+
+fn new_foo_rocks_engine() -> RocksDBEngine {
+    let tmp_root = Builder::new().tempdir().unwrap();
+    let db_path = format!("{}/test", tmp_root.path().display());
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/rocks_engine/test_engine.rs]: hm...这里必须调用下display()吗? 没看文档不确定, 感觉path()后可能已经实现了format的支持.

File: components/epaxos/src/snapshot/rocks_engine/test_engine.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+#[test]
+fn test_base() {
+    let mut eng = new_foo_rocks_engine();
+
+    let none = eng.next_kv(&"init".as_bytes().to_vec(), true);
+    assert_eq!(none, None);
+
+    eng.set_kv("k0".as_bytes().to_vec(), "v0".as_bytes().to_vec())
+        .unwrap();
+    let v0 = eng.get_kv(&"k0".as_bytes().to_vec()).unwrap();
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/rocks_engine/test_engine.rs]: k0 v0这些先let定义出来感觉能少不少代码

File: components/epaxos/src/snapshot/rocks_engine/test_engine.rs (hunk 4 above)
Comment (wenbobuaa) [reply components/epaxos/src/snapshot/rocks_engine/test_engine.rs]: 还得调用下 display，没支持

File: components/epaxos/src/snapshot/rocks_engine/test_engine.rs (hunk 5 above)
Comment (wenbobuaa) [reply components/epaxos/src/snapshot/rocks_engine/test_engine.rs]: 因为 set_kv 定义成传递所有权了，set 之后就没法再引用变量了。这块想个办法重写下..

Comment (drmingdrmer) [reply components/epaxos/src/snapshot/rocks_engine/test_engine.rs]: 用的时候clone下...

File: components/epaxos/src/snapshot/rocks_engine/engine.rs [hunk 6]
```diff
@@ -142,19 +141,16 @@ impl Base for RocksDBEngine {
...
-                    if key == &kv.0 && !include {
-                        iter.next();
-                    } else {
-                        return Some(kv);
-                    }
-                }
-                None => return None,
-            }
+        if !include {
+            seek_key.push('\0' as u8);
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/rocks_engine/engine.rs]: 额...不用这样...这样有额外开销. 先取个next(), 然后看是否include, 如果include并且相等的话,再取一次

Comment (drmingdrmer) [reply components/epaxos/src/snapshot/rocks_engine/engine.rs]: 改key的话还要考虑副作用啥的, 有点不好看˙–˙

Comment (wenbobuaa) [reply components/epaxos/src/snapshot/rocks_engine/engine.rs]: 嗯.. 有理，改一下

File: components/epaxos/src/snapshot/rocks_engine/engine.rs [hunk 7]
```diff
@@ -141,14 +141,26 @@ impl Base for RocksDBEngine {
...
-            seek_key.push('\0' as u8);
+                    if include {
+                        return Some(kv);
+                    };
+                }
+                None => return None,
+            }
+        } else {
+            // TODO may be a rocksdb panic here
+            return None;
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/rocks_engine/engine.rs]: 不valid时先return吧. 能节省13*4个空格.

File: components/epaxos/src/snapshot/rocks_engine/engine.rs [hunk 8]
```diff
@@ -141,14 +141,26 @@ impl Base for RocksDBEngine {
...
+                Some(kv) => {
+                    if &kv.0 != key {
+                        return Some(kv);
+                    };
 
-        if !include {
-            seek_key.push('\0' as u8);
+                    if include {
+                        return Some(kv);
+                    };
```
Comment (drmingdrmer) [inline components/epaxos/src/snapshot/rocks_engine/engine.rs]: 先判断include吧. 要不每次都要比较vector

**Pull request merged**